feat: add Netlify CMS configuration and internationalization support

greatislander opened the pull request:
Adds:

- [x] Netlify CMS configuration for editing posts, pages, site data and site config
- [x] Multilingual support for all collections (including English and Français for demonstration purposes)
- [x] Translations for URL components and other strings (e.g. `/posts/my-post` &rarr; `/articles/mon-article`)
- [x] Localized RSS feeds for each language
- [x] Localization of site head metadata tags (e.g. description)
- [x] Localization of UIO

### Testing

1. Run locally (clone, `npm install`, `npm start`).
2. Verify that internationalized content is editable via http://localhost:8080/admin.
3. Try adding a new language as per the docs.

### Related Issues

- Resolves #48.
- Resolves #141.

jobara commented on src/_includes/layouts/base.njk:
```diff
@@ -30,6 +36,52 @@
+                    enactor: {
+                        "tocTemplate": "/lib/infusion/src/components/tableOfContents/html/TableOfContents.html",
+                        "tocMessage": "/lib/infusion/src/framework/preferences/messages/tableOfContents-enactor.json",
+                        ignoreForToC: {}
```
you should be able to remove this line.

greatislander replied:
@jobara te `ignoreForToc` line?

jobara commented on src/_includes/layouts/base.njk:
```diff
@@ -30,6 +36,52 @@
+        <script>
+        fluid.defaults("fluid.uiOptions.multilingual", {
+            gradeNames: ["fluid.uiOptions"],
+            distributeOptions: {
+                source: "{that}.options.locale",
+                target: "{that messageLoader}.options.resourceOptions.locale"
+            },
```
You could remove this options distributions entirely if you set the `defaultLocale` In the `prefsEditorLoader` options. 

This would replace line 82-83 below.

```javascript
prefsEditorLoader: {
    defaultLocale: "{{ lang | truncate(2, true, '') }}",
    direction: "{{ config.languages | getSiblingValue("code", lang, "dir") }}"
}
```

jobara commented on src/_includes/layouts/base.njk:
```diff
@@ -30,6 +36,52 @@
+                source: "{that}.options.locale",
+                target: "{that messageLoader}.options.resourceOptions.locale"
+            },
+            listeners: {
+                "onPrefsEditorReady.addLanguageAttributesToBody": {
+                    func: "fluid.uiOptions.multilingual.addLanguageAttributesToBody",
+                    args: ["{that}.prefsEditorLoader.prefsEditor.container", "{that}.options.locale", "{that}.options.direction"]
+                }
+            }
+        });
```
Assuming you made the changes from the previous comment to use `defaultLocale`, and also you could remove the `fluid.uiOptions.multilingual.addLanguageAttributesToBody` function.

```suggestion
            listeners: {
                "onPrefsEditorReady.addLanguageAttributesToBody": {
                    this: "{that}.prefsEditorLoader.prefsEditor.container",
                    method: "attr",
                    args: [{
                        "lang": "{that}.options.prefsEditorLoader.defaultLocale",
                        "dir": "{that}.options.prefsEditorLoader.direction"
                    }]
                }
            }
        });
```

You may be able to replace `"{that}.options.prefsEditorLoader"` with `"{prefsEditorLoader}.options"`.

greatislander commented:
@jobara: I'm going to leave the UIO implementation as is, and update the implementation in `eleventy-plugin-fluid` with multilingual support once http://github.com/amb26/infusion/pull/13 makes it in to a dev version I can use.

greatislander commented:
@BlueSlug This is ready for you, I think!

BlueSlug commented on src/collections/pages/pages.11tydata.js:
```diff
@@ -0,0 +1,55 @@
+            return false;
+        },
+        /* Build a permalink using the post title and language key. */
+        permalink: data => {
```
This function is also nearly totally duplicated in `posts.11tydata.js`, so it would be good to reduce duplication where it's practical.

If an author wishes to configure pages and posts separately they will be able to do so deliberately, i.e. we'll encourage a DRY config and if they want to make it WET that'll be up to them.

greatislander replied:
Resolved in 0975e69.

BlueSlug reviewed (changes requested):
Looks pretty good! An error in `pages.11tydata.js` and `posts.11tydata.js` is currently keeping it from running properly, but once that's solved it runs perfectly.

Adding a new locale by following the instructions was straightforward and worked without issue.

greatislander commented on src/collections/pages/pages.11tydata.js:
```diff
@@ -0,0 +1,55 @@
+        /* Build a permalink using the post title and language key. */
+        permalink: data => {
+            /* If this page is a "stub" with no localized title, we assume it does not exist and prevent it from building. */
+            if (!Object.hasOwnProperty(data, "title")) {
```
I meant to use this (but bungled the method):

```
if (!Object.prototype.hasOwnProperty.call(data, "title")) {
```

For reasons enumerated here: https://eslint.org/docs/rules/no-prototype-builtins

What do you think, @BlueSlug?

BlueSlug replied:
This seems like a prudent choice, let's go with what you've suggested!

greatislander commented on src/_includes/partials/global/meta.njk:
```diff
@@ -0,0 +1,7 @@
+    <meta name="description" content="{{ site[lang].description }}">
+    <meta property="og:title" content="{% if title %} {{ title }} | {% endif %} {{ site[lang].name }}">
```
Created a documentation issue for this: https://github.com/fluid-project/trivet/issues/154

greatislander commented on README.md:
```diff
@@ -41,13 +46,47 @@ directory to the web root of your server.
+
+2. Add sections for the new language to [`src/_data/site.json`](src/_data/site.json) and [`src/_data/translations.json`](src/_data/translations.json),
+   translating the content from the English source. The key will always be the language code as used in `config.json`.
+3. Add the language to the [locales configuration](https://github.com/fluid-project/trivet/blob/b899db1/src/admin/config.yml#L13)
```
@BlueSlug: in 54db27d I've set up something else we talked about— the Netlify CMS config file is now templated to include the languages enumerated in `config.json` and use the first language as the default in its `locales` block. So I was able to remove this step altogether! Adding a language to `config.json` will now automatically update the CMS.

BlueSlug replied:
Awesome!

greatislander commented on src/collections/pages/pages.11tydata.js:
```diff
@@ -0,0 +1,55 @@
+    layout: "layouts/page.njk",
+    eleventyComputed: {
+        /* Determine the language of this item based on the language code in the file path. */
+        lang: data => data.page.filePathStem.replace(/\/collections\/pages\/([A-Za-z-]*)\/[a-z0-9-]*/g, "$1"),
```
Resolved in 62a9c5e.

amb26 commented on .eleventy.js:
```diff
@@ -34,25 +34,39 @@ module.exports = function (config) {
+    siteConfig.languages.forEach(lang => {
+        config.addCollection(`posts_${lang.code}`, collection => {
+            return [
+                ...collection.getFilteredByGlob(`./src/collections/posts/${lang.code}/*.md`).filter(livePosts)
```
Any particular reason to use the array + spread syntax here rather than returning the original array value?

greatislander replied:
It was in the original code that I picked up from our GSOC contributor and I never thought to change it. And you're right, it's not necessary. I thought at first that `getFilteredByGlob` might be returning a Set rather than an array but I checked and that's not the case.

BlueSlug commented on src/collections/pages/pages.11tydata.js:
```diff
@@ -0,0 +1,36 @@
+                    parent: data.lang,
+                    order: data.order,
+                    /* If a key is set, use that for the navigation item label; otherwise use the page title. */
+                    key: data.hasOwnProperty("key") ? data.key : data.title
```
```suggestion
                    key: Object.prototype.hasOwnProperty.call(data, "key") ? data.key : data.title
```

amb26 commented on .eleventy.js:
```diff
@@ -34,25 +34,37 @@ module.exports = function (config) {
     config.addPassthroughCopy({"src/posts/images": "posts/images"});
 
+    config.addNunjucksFilter("getSiblingValue", function (value, inputKey, inputValue, siblingKey) {
+        let siblingValue = false;
```
Clearer here is 

let matchedItem = value.find(item => item[inputKey] === inputValue);
return matchedItem && matchedItem[siblingKey];

You might appreciate some JSDocs here in case you forget someday what this function does - terribly easy to do!

BlueSlug replied:
Indeed, JSDoc throughout the project would be excellent! We could file an issue for that and tackle it separately, no need to add everything with this PR.

BlueSlug reviewed (commented):
Pretty much everything looks great! I noticed that the Posts page doesn't have a Language link in the navigation, is that intentional?

greatislander commented:
@BlueSlug It should have a link! I'll take a look first thing tomorrow.

BlueSlug reviewed (approved):
Looks good to me, I'm comfortable merging this.

@jobara & @amb26: do you have any further thoughts or comments you'd like to add?

jobara commented on src/_data/config.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "maxPostsInFeed": 10,
+  "languages": [
+    {
+      "code": "en-CA",
+      "dir": "ltr",
+      "name": "English"
+    },
+    {
+      "code": "fr-CA",
+      "dir": "ltr",
+      "name": "Français"
+    }
+  ],
```
Is there a specific reason to define this as an array instead of an object? If it was an object you could key off of the language code, as you do in other places like site.json and translations.json. Also you probably wouldn't need `getSiblingValue` anymore.

greatislander replied:
Yes— this makes it editable in the CMS.

jobara replied:
Can you expand on this a bit more. What's the restriction? Does the CMS consume the config.json file directly and requires an array?

jobara commented on src/admin/config.njk:
```diff
@@ -0,0 +1,86 @@
+i18n:
+  structure: multiple_folders
+  locales: [{% for language in config.languages %}{{language.code}}{% if loop.last !== true %}, {% endif %}{% endfor %}]
+  default_locale: {{ config.languages[0].code }}
```
It appears that you are specifying the default language based on the first item in the languages array. However, in feed.njk it appears to explicitly look for "en-CA". It would probably be better to have the default language set in the configuration and just check for that wherever you need to make some action based on the default locale.

greatislander replied:
@BlueSlug and I discussed this and felt that the first language should be the default rather than adding another config item for the default language. I should update the feed to use the first language as default, though.

BlueSlug replied:
I'm on the fence now, actually. A separate config item would mean a bit of duplication, though it would make it fairly explicit which is the default which *could* be easier for an author to fiddle with.

greatislander replied:
The problem is that Netlify CMS can't dynamically update the list of languages that a user has added to allow one to be selected as default. SO if you need to add a new language and then change the default language, you'd need to:

1. Add the new language.
2. Publish.
3. Wait for the site to build.
4. Reload the CMS.
5. Select the new default language from the list.
6. Publish.

I do think that having the first language be the default makes sense and is less convoluted. Perhaps if we add a hint to the CMS which says "the first language in this list will be the site's default language— reorder the list if you want to make the default a different language".

jobara replied:
I'm a bit confused. Can you add new languages and reorder the list of them through the CMS directly? What is the difference that allows reordering languages but not selecting a default when adding a new language?

jobara commented on src/collections/pages/fr-CA/about.md:
```diff
@@ -0,0 +1,8 @@
+order: 2
+excerpt: "Ceci est une brève description de la page. "
+---
+Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore
+magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
+gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.
```
Similar to other example pages, the about page should have a real description and be localized to the specified language.

BlueSlug replied:
<insert "It's all Greek to me" joke>

jobara commented on src/transforms/html-min-transform.js:
```diff
@@ -14,7 +14,7 @@ https://github.com/fluid-project/trivet/raw/main/LICENSE.md.
 
 module.exports = function htmlMinTransform(value, outputPath) {
-    if (outputPath.indexOf(".html") > -1) {
+    if (outputPath && outputPath.includes(".html")) {
```
Probably doesn't matter but might be more appropriate/safer to use [`endsWith`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith). That way you just check the end of the string, instead of seeing if `.html` appears anywhere in the `outputPath`.

jobara commented on src/utils/getShortLang.js:
```diff
@@ -0,0 +1,5 @@
+"use strict";
+
+module.exports = lang => {
+    return lang.split("-")[0];
+};
```
Don't know how useful this utility is. It is only used once, and in feed.njk the split is just done directly.

jobara commented on src/utils/generatePermalink.js:
```diff
@@ -0,0 +1,27 @@
+
+    if (collectionType === "pages") {
+        if (data.page.fileSlug === lang) {
+            return (lang === "en-CA") ? "/" : `/${shortLang}/`;
+        }
+
+        return (lang === "en-CA") ? `/${slug}/` : `/${shortLang}/${slug}/`;
```
As mentioned before, it's probably best to move the default language to the config and check against that value instead of the hardcoded `"en-CA"`.

jobara commented on src/utils/getLang.js:
```diff
@@ -0,0 +1,5 @@
+"use strict";
+
+module.exports = (filePathStem, collection) => {
+    return filePathStem.replace(new RegExp(`\/collections\/${collection}\/([A-Za-z-]*)\/[a-z0-9-]*`, "g"), "$1");
```
Is the `/collections/` path segment an 11ty convention or requirement?

greatislander replied:
It's a choice I made for Trivet. I didn't have a `collections` directory for weavly.org and the `src` directory is much more cluttered as a result: https://github.com/codelearncreate/weavly.org/tree/main/src

jobara replied:
Hmm.. I was going to suggest that you parameterize the base path; however, in skimming through the code again, it would seem that someone would have to have to rewrite quite a bit to use a different directory name/structure. I suppose this makes `collections` a requirement for using trivet. As long as that's clear in the docs, I guess this is fine.

jobara commented on src/utils/generatePermalink.js:
```diff
@@ -0,0 +1,27 @@
+        return false;
+    }
+
+    const lang = getLang(data.page.filePathStem, collectionType);
+    const shortLang = lang.split("-")[0];
+    const slug = slugFilter(data.title);
+
+    if (collectionType === "pages") {
+        if (data.page.fileSlug === lang) {
+            return (lang === config.languages[0].code) ? "/" : `/${shortLang}/`;
+        }
+
+        return (lang === config.languages[0].code) ? `/${slug}/` : `/${shortLang}/${slug}/`;
+    } else {
+        return (lang === config.languages[0].code) ? `/${translations[lang][collectionType]}/${slug}/` : `/${shortLang}/${translations[lang][collectionType]}/${slug}/`;
+    }
+};
```
What would happen here if you have translations for multiple locales? e.g. en-CA, en-UK, en-US? I'd suppose that if en-CA is the default that you'd get that from the default URL. However, the en-UK and en-US would probably override/merge for the /en/ paths.

greatislander replied:
@jobara True— I guess my expectation is that as a starter kit this is a good approach that can be modified for more complex cases where a bunch of national dialects of the same language are being used. In that instance someone could use the full code in the permalink instead of the `shortLang` with minimal changes. The goal here was to make the URL as simple as possible (`/fr/articles` instead of `/fr-CA/articles`) but still be able to add the HTML `lang` attribute in a way that reflects the full culture code (language plus country).

jobara replied:
An alternative would be to check if there are multiples of the same language with just different locales. In that way you could use the short-code if the language is unique and the full locale if not. This would be more complex in the code though.

In general I guess you need to balance the simplicity of the code, the url, and potential user issues. If you leave as is, it should at least be documented, as I'm not sure how obvious it would be. In particular, if the language can be added via the CMS directly, I'm not sure that the maintainer would be able to debug this issue.

greatislander replied:
> An alternative would be to check if there are multiples of the same language with just different locales. In that way you could use the short-code if the language is unique and the full locale if not. This would be more complex in the code though.

I've implemented this in 53903f5.

jobara commented on src/utils/getLanguageSlug.js:
```diff
@@ -0,0 +1,22 @@
+    /* Trim the country code from the end of the language. */
+    const shortLang = lang.split("-")[0];
+
+    /* Create array of all supported language codes with the country code trimmed. */
+    const codes = config.languages.reduce((accumulator, currentValue) => {
+        accumulator.push(currentValue.code.split("-")[0]);
+        return accumulator;
+    }, []);
```
I think what you are doing here is to take an array like 

```javascript
[{code: "en-CA"}, {code: "en-GB"}, {code: "en-US"}, {code: "fr-CA"}, {code: "fr-FR"}, {code: "es-MX"}]` 
```

and convert it to an array like:

```javascript
["en", "en", "en", "fr", "fr", "es"]
```

If so, you could simply the code by using [Array.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) instead. However, you could continue to use reduce and instead of outputting an array, output an Object of counts. That is with the key the trimmed lang and the value the number of occurrences. Such as:

```json
{
    "en": 3,
    "fr": 2,
    "es" 1
}
```

Something like the following should accomplish that. 

```javascript
const codes = config.languages.reduce((accumulator, currentValue) => {
    let lang = currentValue.code.split("-")[0];
    accumulator[lang] = (++accumulator[lang] || 1);
    return accumulator;
}, {});
```

greatislander replied:
I like this alternate usage of the reducer— I'm going to try it.

amb26 replied:
A bit sleazy having the conditional double side-effect there - one might want to say
````
accumulator[lang] = (accumulator[lang] || 0) + 1; 
````
to be a touch clearer - note that in the bad old days (C++) such expressions were actually undefined!
But it's not really clear why the counts are actually required - isn't it enough to simply say
````
accumulator[lang] = true
````
and then emit an Object.keys() of the result?

Note that the original method is also fine, after passing the result through a Set comprehension as in this answer: https://stackoverflow.com/a/33121880

greatislander replied:
> But it's not really clear why the counts are actually required - isn't it enough to simply say
> 
> ```
> accumulator[lang] = true
> ```
> 
> and then emit an Object.keys() of the result?

Not really! The intent here is to find languages with more than one national/regional variant so that slugs don't overwrite each other.

For example, let's say we have a site with `en-CA` and `fr-CA` as in the example config. It's cleaner to have URLs constructed like this: `/en/something` or `/fr/quelque-chose` so we want to trim the country code off the language.

However, imagine a more complex Trivet-based site which has many variations: `es-ES`, `es-MX`, `pt-BR`, `pt-PT`. If we were to follow the same logic and trim the country code that would break the site because both the Spanish (Spain) and Spanish (Mexico) languages would end up with the `/es` language slug.

So by building an object of short language slugs ([ISO 639-1 codes](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes)) with the count for each, we can use the short code if the count is 1 but use the long codes for languages with a count > 2. Does this make sense, @amb26? I will make the change you suggest related to the conditional double side-effect.

jobara replied:
@amb26 I think the issue is that you need to know when there are more than 1. If we just added a boolean flag or used a set, we'd know that there was an occurrence of a language, but not the number of locales present. The number of locales is used to determine if the url can be simplified to something like `/fr/post/` or requires the longer form like `/fr-CA/posts/` and `/fr-FR/posts/`.

jobara commented:
@greatislander would you be able to expand a bit on how the languages are added to the site. I had some additional questions  about the requirements and etc. For example, does the CMS provide an interface for adding the languages, or does this happen by someone directly committing the updates to the GitHub repo, for example?

greatislander commented:
> @greatislander would you be able to expand a bit on how the languages are added to the site. I had some additional questions about the requirements and etc. For example, does the CMS provide an interface for adding the languages, or does this happen by someone directly committing the updates to the GitHub repo, for example?

A lot of this is covered here: https://github.com/fluid-project/trivet/blob/53903f56eb5265e92f034dd4e8fac47b892bee8c/README.md#internationalization

The CMS does provide an interface to add the languages (I'm not sure if you tested that part) but it does still need some manual intervention in GitHub, namely steps 2 and 3.

jobara commented:
> A lot of this is covered here: https://github.com/fluid-project/trivet/blob/53903f56eb5265e92f034dd4e8fac47b892bee8c/README.md#internationalization
> 
> The CMS does provide an interface to add the languages (I'm not sure if you tested that part) but it does still need some manual intervention in GitHub, namely steps 2 and 3.

I wasn't sure how to access the CMS to test. However,  I now see that the [local_backend](https://www.netlifycms.org/docs/beta-features/#working-with-a-local-git-repository) option was enabled and found the `/admin` URL to access it locally.

I did read through that part of the README, but it sounded like steps 1 - 3 would happen in GitHub and step 4 on the CMS. It isn't clear from the README that step 1 can also be performed through the CMS. Based on your comments, it seems that step 2 and 3 require manual intervention in GitHub. It isn't clear from the configuration UI that there are additional tasks required, and that they must happen outside of the CMS admin interface. 

I understand things a bit better now. You're using the [list widget](https://www.netlifycms.org/docs/widgets/list/) which allows for outputting the current supported languages, add additional ones, and reorder them. One thing I couldn't figure out how to do, was to reorder the list with the keyboard. If that isn't possible, keyboard only users wouldn't be able to change the default language through the CMS.

If in fact you need to have manual intervention through GitHub to fully configure a new language, and you can't change the default value via the CMS with keyboard only, I'm wondering if adding a new language should be restricted to direct GitHub commits. At least, until a fully featured option is available through the CMS. If you were to change the default language selection to a select or possibly other widget, it should be possible to still change the default langue through the CMS. Although it would be restricted to languages that are already available to the system. There may be some other combinations or widgets, interactions, etc. that could be employed to make all this work as well.

jobara commented on src/_data/config.json:
```diff
@@ -10,6 +10,11 @@
+    {
+      "code": "fr-FR",
+      "dir": "ltr",
+      "name": "Français"
```
The double Français in the list of languages is a bit confusing to look at. I'm guessing you can't actually require much here, but it might be better to have as an example `Français (fr-FR)` and `Français (fr-CA)`. Or some other way to easily distinguish them.

<img width="836" alt="Screen Shot 2021-07-19 at 10 03 34 AM" src="https://user-images.githubusercontent.com/556024/126177524-5934c481-0b25-42c9-9658-ed8c46e1bea4.png">

jobara commented on src/utils/getLanguageSlug.js:
```diff
@@ -8,12 +8,13 @@ module.exports = (lang) => {
     const codes = config.languages.reduce((accumulator, currentValue) => {
-        accumulator.push(currentValue.code.split("-")[0]);
+        let lang = currentValue.code.split("-")[0];
+        accumulator[lang] = (++accumulator[lang] || 1);
```
@amb26's suggestion to use `accumulator[lang] = (accumulator[lang] || 0) + 1;` would be clearer here. 

See: [comment](https://github.com/fluid-project/trivet/pull/148#discussion_r670799553).

BlueSlug commented:
> > A lot of this is covered here: https://github.com/fluid-project/trivet/blob/53903f56eb5265e92f034dd4e8fac47b892bee8c/README.md#internationalization
> > The CMS does provide an interface to add the languages (I'm not sure if you tested that part) but it does still need some manual intervention in GitHub, namely steps 2 and 3.
> 
> I wasn't sure how to access the CMS to test. However, I now see that the [local_backend](https://www.netlifycms.org/docs/beta-features/#working-with-a-local-git-repository) option was enabled and found the `/admin` URL to access it locally.
> 
> I did read through that part of the README, but it sounded like steps 1 - 3 would happen in GitHub and step 4 on the CMS. It isn't clear from the README that step 1 can also be performed through the CMS. Based on your comments, it seems that step 2 and 3 require manual intervention in GitHub. It isn't clear from the configuration UI that there are additional tasks required, and that they must happen outside of the CMS admin interface.
> 
> I understand things a bit better now. You're using the [list widget](https://www.netlifycms.org/docs/widgets/list/) which allows for outputting the current supported languages, add additional ones, and reorder them. One thing I couldn't figure out how to do, was to reorder the list with the keyboard. If that isn't possible, keyboard only users wouldn't be able to change the default language through the CMS.
> 
> If in fact you need to have manual intervention through GitHub to fully configure a new language, and you can't change the default value via the CMS with keyboard only, I'm wondering if adding a new language should be restricted to direct GitHub commits. At least, until a fully featured option is available through the CMS. If you were to change the default language selection to a select or possibly other widget, it should be possible to still change the default langue through the CMS. Although it would be restricted to languages that are already available to the system. There may be some other combinations or widgets, interactions, etc. that could be employed to make all this work as well.

It's worth mentioning at this point that, given this is intended as a website starter kit and not a full-fledged solution, a certain amount of programming knowledge is assumed, and indeed required in order to get anything set up beyond the sample pages. With that in mind, I think it's ok if the default language can only be set or new languages can only be added via the config files rather than from within Netlify CMS.

greatislander commented:
> It's worth mentioning at this point that, given this is intended as a website starter kit and not a full-fledged solution, a certain amount of programming knowledge is assumed, and indeed required in order to get anything set up beyond the sample pages. With that in mind, I think it's ok if the default language can only be set or new languages can only be added via the config files rather than from within Netlify CMS.

Thanks, @BlueSlug — yes, this reflects my thinking. I've removed the ability to manage site languages from the CMS. I've also updated the docs (d648ebb) to reflect that the default language will be the first that appears in `src/_data/config.json`.

jobara commented:
@greatislander and @BlueSlug with the language option removed from the CMS, it may be worth thinking about converting the languages option in the config.json from an array to an object. I think this will simplify the code, e.g. removing `getSiblingValue`. It will also provide an explicit declaration for the default language; which may be easier to set than re-arranging an array of languages. This would also provide the opportunity to throw a build time error if the default language code doesn't match any languages.

I don't think this will end up being something that gets changed much after the initial setup, so it might be more worthwhile to focus on what is easier to maintain for an integrator/user of trivet as they go about working on a site that they start from it.

jobara commented on src/_includes/layouts/base.njk:
```diff
@@ -30,6 +36,51 @@
+                    }
+                }
+            },
+            locale: "{{ lang | truncate(2, true, '') }}",
```
Not sure I'm understanding this correctly. It looks like it will convert something like "fr-CA" to "fr", and pass that in as a locale. However, if there are multiple locales such as "fr-CA" and "fr-FR" it may not set the correct locale for UIO's message bundles. Perhaps you're doing this to reduce requests for UIO's template fetching for fallbacks?

Is there a way for the site builder (user of trivet) to add translations for UIO themselves? It's worth updating the README to mention how to do this. If it isn't possible, the README should probably mention that too, and also note which languages are available for UIO, maybe linking to the Infusion repo if appropriate.

greatislander replied:
> Not sure I'm understanding this correctly. It looks like it will convert something like "fr-CA" to "fr", and pass that in as a locale. However, if there are multiple locales such as "fr-CA" and "fr-FR" it may not set the correct locale for UIO's message bundles. Perhaps you're doing this to reduce requests for UIO's template fetching for fallbacks?

Pretty much! As far as I can tell, UIO's only message bundles for country-specific languages are `en_US`, `en_CA` and `pt_BR`. In the interests of simplicity I opted to use the short (`en`) bundle for English which will allow Trivet and UIO to properly match up locales for all the other supported ones… but the Brazilian Portuguese locale makes this not as effective as I'd hoped it would be. 

> Is there a way for the site builder (user of trivet) to add translations for UIO themselves? It's worth updating the README to mention how to do this. If it isn't possible, the README should probably mention that too, and also note which languages are available for UIO, maybe linking to the Infusion repo if appropriate.

I think it would be much better for the Fluid ecosystem if site builders were encouraged to add translations to UIO directly so all other UIO integrators can benefit from them. Theoretically it would be possible for site builder to initialize UIO with a different messages path and copy all the translations they needed (and add some of their own) but I think the best bet is to indicate which locales are available in the README as per your suggestion.

greatislander replied:
I've attempted to resolve this and eliminate the need for all that permalink determination logic by expanding the languages config block slightly in d2d694b. It now looks like this:

https://github.com/greatislander/trivet/blob/d2d694b97bdd19336f3f31d2b8e11b1f59d0a20a/README.md#L79-L88

jobara replied:
> Pretty much! As far as I can tell, UIO's only message bundles for country-specific languages are `en_US`, `en_CA` and `pt_BR`. In the interests of simplicity I opted to use the short (`en`) bundle for English which will allow Trivet and UIO to properly match up locales for all the other supported ones… but the Brazilian Portuguese locale makes this not as effective as I'd hoped it would be. I think at minimum I need to change my approach so that the replacement logic is to truncate OR change from kebab case to snake case depending on the locale in question. It's a bit messy but the problem with _not_ doing that is that in the example included in this repo, we have:
> 
> * en-CA
> * fr-CA
> 
> UIO includes:
> 
> * en_CA
> * fr
> 
> So I need to truncate one and replace the hyphen with an underscore in the other…

For starters we probably should have a fallback for Portugese if the region code isn't added. I've filed this as [FLUID-6675](https://issues.fluidproject.org/browse/FLUID-6675).

For "fr", if you pass in "fr_CA" it will fallback to "fr" so you don't actually have to truncate it. However, it would trigger additional requests to look for the "fr_CA" localization first. Considering where all this code runs, it would probably be better to truncate, or just add the "fr_CA" translation to UIO in Infusion. It probably already is "fr-CA" anyways, but @BlueSlug would be better to confirm that.

I can't remember why we use the "en_CA" vs "en-CA" format. Perhaps @amb26 would remember. Feel free to file a JIRA if you think that should be changed. It's something we could consider.

> I think it would be much better for the Fluid ecosystem if site builders were encouraged to add translations to UIO directly so all other UIO integrators can benefit from them. Theoretically it would be possible for site builder to initialize UIO with a different messages path and copy all the translations they needed (and add some of their own) but I think the best bet is to indicate which locales are available in the README as per your suggestion.

Yes, I agree adding more translations to UIO directly would be ideal. Although it's great to hear that it's possible for site builder to configure UIO directly if needed. I think we'll eventually want to add [crowdin](https://crowdin.com) or something like that for UIO, if possible. Although I'm going to guess this would have to wait till the Infusion 5.x timeline ( could be before the end of the year ). Until then, we'll likely be constrained to what's currently available.

greatislander replied:
Is the `uioSlug` solution reasonable for now, do you think? It makes the mapping explicit which will save additional lookups/fallback logic.

BlueSlug replied:
> For "fr", if you pass in "fr_CA" it will fallback to "fr" so you don't actually have to truncate it. However, it would trigger additional requests to look for the "fr_CA" localization first. Considering where all this code runs, it would probably be better to truncate, or just add the "fr_CA" translation to UIO in Infusion. It probably already is "fr-CA" anyways, but @BlueSlug would be better to confirm that.

IETF / BCP-47 Language Tags use a hyphen to separate the language part from the locale part, so it makes sense to be consistent with that specification. I can't recall why UIO/Infusion uses underscores, unfortunately.

amb26 replied:
UIO went with underscores to be consistent with the Java locale format https://www.oracle.com/technical-resources/articles/javase/locale.html which was the common format for a lot of resource bundles at the time Infusion was being designed. We should switch to BCP-47 when we have the Infusion 5 rewrite.

jobara replied:
I've filed this as [FLUID-6676](https://issues.fluidproject.org/browse/FLUID-6676).

jobara commented on src/admin/config.njk:
```diff
@@ -0,0 +1,70 @@
+  preview: false
+i18n:
+  structure: multiple_folders
+  locales: [{% for code, language in config.languages %}{{code}}{% if loop.last !== true %}, {% endif %}{% endfor %}]
```
Is the `language` variable needed here?

greatislander replied:
Yes— this is the only downside to the change in `config.languages` structure. When [iterating over an object](https://mozilla.github.io/nunjucks/templating.html#for), Nunjucks requires the following syntax:

```njk
{% for key, value in object %}
{# Do something… #}
{% endfor %}
```

All that's needed here is the key but I'm pretty sure I can't do this:

```njk
{% for key in Object.keys(object) %}
{# Do something… #}
{% endfor %}
```

So the `language` variable is needed.

jobara replied:
Thanks for the link to the nunjucks docs. Probably not worth changing to the second format even if it is possible.

jobara commented on src/posts.fr-CA.md:
```diff
@@ -0,0 +1,15 @@
+  parent: fr-CA
+  key: Articles
+  order: 3
+permalink: "/fr/articles/{% if pagination.pageNumber > 0 %}page/{{ pagination.pageNumber + 1 }}/{% endif %}"
```
is the `fr` in the path going to be an issue if "fr-CA" is used in the URLs instead?

greatislander replied:
It would be, but if the site builder was adding another French locale they would be creating another posts file (e.g. `posts.fr-FR.md`) and would be setting the permalinks in each.

jobara replied:
I agree that the fix for a site build is easy, but I worry that the debugging of this issue might be hard. Is there a way to surface this/draw attention to it. Even if it's just a note in the README or something.

greatislander replied:
In 7309c82, I changed the permalink so it uses the `config.languages[lang].slug` value instead of a hard-coded `fr`. This should better guidance for adding other languages.

jobara commented on src/posts.fr-CA.md:
```diff
@@ -0,0 +1,15 @@
+  data: collections.posts_fr-CA
+  size: 10
+  alias: posts
+---
```
There seems to be a lot of duplication between src/posts.fr-CA.md and src/posts.en-CA.md. I'd assume additional languages would also be similar. It may make sense to template these files as well at some point.

greatislander replied:
I can look into this. Because the URL structure differs between the default language (no language slug in front) and other languages, there's a bit of differentiation that's needed so I had thought it would make sense to not try to template these at the risk of making it confusing for a site builder to figure out where all the post archive pages are coming from. But I will see what I can do.

jobara replied:
Just to clarify, I didn't mean this needed to happen in this PR, but something to think about for the future. If you think it make sense to do now, go for it, but we can also punt it for another time.

greatislander replied:
Probably makes sense to punt it— it would open up a whole can of worms…

jobara commented on .eleventy.js:
```diff
@@ -42,17 +49,17 @@ module.exports = function (config) {
-        return [...collection.getFilteredByGlob("./src/posts/*.md").filter(livePosts)]
-            .reverse()
-            .slice(0, site.maxPostsInFeed);
+    Object.keys(siteConfig.languages).forEach(lang => {
```
Probably somewhere before this there should be some validation that the `defaultLanguage` exists in the set of configured languages, and throw an appropriate error if not.

greatislander replied:
I've added an error to the output to explain other build problems that might arise if this is misconfigured.

jobara commented on .eleventy.js:
```diff
@@ -84,6 +78,12 @@ module.exports = function (config) {
     });
 
+    config.on("beforeBuild", () => {
+        if (!Object.keys(siteConfig.languages).includes(siteConfig.defaultLanguage)) {
+            console.error("The default language configured in src/_data/config.json is not one of your site's supported languages."); // eslint-disable-line no-console
+        }
```
I don't think you need to convert the `siteConfig.languages` Object to an array of keys. You could probably get away with just checking if the value is truthy. 

Something like the following might work:

```javascript
if (!siteConfig.langauges[siteConfig.defaultLanguage]) {
    throw new Error(`The default language, ${siteConfig.defaultLanguage}, configured in src/_data/config.json is not one of your site's supported languages.`);
}
```
I'm suggesting throwing an error so that there is a build time error, rather than outputting the site in a potentially invalid state. There may be another way to achieve this with 11ty.

greatislander replied:
Gotcha— this works! Added in c5b7b64.

jobara commented on README.md:
```diff
@@ -67,13 +67,20 @@ following changes need to be made:
+
+   * an optional short form of the language code for use in permalinks
+   * an optional short form of the language code that corresponds to an available message bundle locale for Infusion's
+     [User Interface Options](https://github.com/fluid-project/infusion/tree/main/src/framework/preferences/messages)
```
This is unfortunate, but maybe we can clean this up later. Possibly after changes to UIO in Infusion.

jobara commented on README.md:
```diff
@@ -67,13 +67,20 @@ following changes need to be made:
-   (endonym) as the object values:
+   as the key, and add the following object values:
+
+   * an optional short form of the language code for use in permalinks
+   * an optional short form of the language code that corresponds to an available message bundle locale for Infusion's
+     [User Interface Options](https://github.com/fluid-project/infusion/tree/main/src/framework/preferences/messages)
```
Also are these actually optional? I did see one case where there was a default value, but wasn't sure about the rest.

greatislander replied:
Oops, no, they aren't optional... I was going to try to do that but it didn't work and I forgot to update the doc. Fixed in 757a2c2.

jobara commented on .eleventy.js:
```diff
@@ -77,6 +78,12 @@ module.exports = function (config) {
         }
     });
 
+    config.on("beforeBuild", () => {
+        if (!siteConfig.languages[siteConfig.defaultLanguage]) {
+            throw new Error(`The default language, ${siteConfig.defaultLanguage}, configured in src/_data/config.json is not one of your site's supported languages.`);
+        }
+    });
```
I tried triggering this error, but it gets swallowed up. Basically the error is initially reported to the terminal, but gets removed when the webpack stuff starts up. 

_**Note**: this is only an issue with `npm start`, `npm run build` reports the errors._

It seems this is caused by: 

1. 11ty still exiting with a 0 exit code, despite the error thrown. 
2. concurrently still running other concurrent processes when one failed.

To address 1, you can add `process.exitCode = 1;` just before line 83 where the error is thrown. This tells the process to set the exit code to 1. You may want to see if another [Exit Code](https://nodejs.org/api/process.html#process_exit_codes) makes more sense. Possibly 9, but whatever seems most appropriate.

To Address 2, you can set the `--kill-others-on-fail` flag for concurrently. This tells it to end all of the other process if one exits with a non-zero exit code.

greatislander replied:
Resolved in 3b17bb6.

jobara commented on package.json:
```diff
@@ -7,11 +7,12 @@
         "start": "npm-run-all -l clean -p start:*",
         "start:webpack": "mix watch",
-        "start:eleventy": "cross-env NODE_ENV=development eleventy --serve"
+        "start:eleventy": "concurrently \"npm run dev\" \"netlify-cms-proxy-server\""
```
Is there a need for both concurrently and npm-run-all? Would it be possible to just use one of these, or do they provide unique features for specific cases?

greatislander replied:
It looks like npm-run-all with the parallel flag (or [run-p](https://github.com/mysticatea/npm-run-all/blob/master/docs/run-p.md)) can handle parallelized commands in the same way that concurrently does. I've removed concurrently in favour of run-p in 3b17bb6.

jobara commented on src/posts.fr-CA.md:
```diff
@@ -0,0 +1,15 @@
+  parent: fr-CA
+  key: Articles
+  order: 3
+permalink: "/{{ config.languages[lang].slug }}/articles/{% if pagination.pageNumber > 0 %}page/{{ pagination.pageNumber + 1 }}/{% endif %}"
```
I assume this issue is related to this line, but it might be something else. If you set the site to default to `fr-CA`. The URLs for Home and About have `/en/` in the path for the English pages, but Posts does not. Conversely, the French pages for Home and About do not have `/fr/` in the URL but the Posts page does. 

Navigating from the Posts page **does not** break the language selection.

greatislander replied:
Fixed in 2150a10 so that changing defaultLanguage updates these permalinks appropriately.

jobara commented:
@BlueSlug and @greatislander I've wrapped up my review and approved the PR. I'll pass it back to you in case there's anything else.

BlueSlug commented on src/_includes/partials/components/brand.njk:
```diff
@@ -1,3 +1,7 @@
+                {% if lang !== config.defaultLanguage %}
+                {% set homeUrl = "/" + config.languages[lang].slug %}
+                {% endif %}
+                <a rel="home" href="{{ homeUrl }}"{% if page.url === homeUrl %} aria-current="page"{% endif %}>Trivet</a>
```
It would make more sense if the link text was set to `site[lang].name`, i.e. the site name set in `site.json` which is also used in the page `title`

greatislander replied:
@BlueSlug fixed! I had to set the `lang` to `config.defaultLang` when undefined to work around the 404 issue.

BlueSlug commented on src/_includes/partials/components/brand.njk:
```diff
@@ -1,3 +1,7 @@
             <div class="brand">
-                <a rel="home" href="/"{% if page.url === '/' %} aria-current="page"{% endif %}>Trivet</a>
+                {% set homeUrl = "/" %}
+                {% if lang !== config.defaultLanguage %}
```
As we discussed on a call, if the language is not the default, the homepage URL isn't defined, so the homepage link on 404 pages will redirect the user to `/undefined`, which of course returns the 404 page, and so on... :)

```suggestion
                {% if lang and lang !== config.defaultLanguage %}
```

BlueSlug reviewed (changes requested):
Everything is looking great, and the instructions to add a new language are still working, despite my best efforts :) Just a couple small issues with the brand component, so once those are resolved I'll give my approval.

It would also be good to file an issue to explore localized 404 pages, given the fact that Netlify will be handling 404 responses statically; this isn't something that I feel is necessary to resolve in this PR.
